From 4b0442a210ce307a5995811831d58046fe84455b Mon Sep 17 00:00:00 2001
From: Bharath <linux.bhar@gmail.com>
Date: Thu, 4 Jul 2019 05:27:44 -0600
Subject: [PATCH 2/6] Changed kernel version string

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 26c92f8..3f4f627 100644
--- a/Makefile
+++ b/Makefile
@@ -2,7 +2,7 @@
 VERSION = 5
 PATCHLEVEL = 1
 SUBLEVEL = 0
-EXTRAVERSION =
+EXTRAVERSION =predm
 NAME = Shy Crocodile
 
 # *DOCUMENTATION*
-- 
1.8.3.1

